pkg/planner/cascades/rule/ruleset/rule_set.go

```diff
@@ -15,14 +15,70 @@
 package ruleset
 
 import (
+	"github.com/bits-and-blooms/bitset"
+	"github.com/pingcap/tidb/pkg/planner/cascades/memo"
 	"github.com/pingcap/tidb/pkg/planner/cascades/pattern"
 	"github.com/pingcap/tidb/pkg/planner/cascades/rule"
-	"github.com/pingcap/tidb/pkg/planner/cascades/rule/apply/decorrelate_apply"
+	"github.com/pingcap/tidb/pkg/planner/cascades/rule/apply/decorrelateapply"
+	"github.com/pingcap/tidb/pkg/planner/core/operator/logicalop"
 )
 
-// DefaultRuleSet is default set of a series of rules.
-var DefaultRuleSet = map[pattern.Operand][]rule.Rule{
-	pattern.OperandApply: {
-		decorrelateapply.NewXFDeCorrelateApply(),
+// SetType is the type of rule set.
+type SetType uint
+
+const (
+	// DefaultNone indicates this is none rule.
+	DefaultNone SetType = iota
+	// XFSetDeCorrelateApply indicates a set of sub-rules related to de-correlate an apply.
+	XFSetDeCorrelateApply
+)
+
+// DefaultRuleSets indicates the all rule set.
+var DefaultRuleSets = map[pattern.Operand]*OperandRules{
+	pattern.OperandApply: OperandApplyRules,
+}
+
+// OperandRules wrapper all the rules rooted from one specified operator.
+type OperandRules struct {
+	setMap  map[SetType][]rule.Rule
+	setList []rule.Rule
+}
+
+// ListRules is a list of rules.
+type ListRules []rule.Rule
+
+// Filter mask out rules which is in mask uint64.
+func (l ListRules) Filter(mask *bitset.BitSet) ListRules {
+	res := make([]rule.Rule, 0, len(l))
+	for _, one := range l {
+		if mask.Test(one.ID()) {
+			res = append(res, one)
+		}
+	}
+	return res
+}
+
+// Filter return the specified operand rule list filter with operator's attribute.
+func (ors *OperandRules) Filter(ge *memo.GroupExpression) ListRules {
+	// special case for de-correlate apply.
+	// for intermediate apply, we should only apply de-correlate rule for short path.
+	if ge.GetBaseLogicalPlan().(*logicalop.BaseLogicalPlan).Flag&logicalop.ApplyGenFromXFDeCorrelateRuleFlag > 0 {
+		return ors.setMap[XFSetDeCorrelateApply]
+	}
```

Comment:
this needs a clear example

Reply:
yes, test is mock case, the later pull request will add flag for intermediary apply